### DISCUSSION TITLE
Question on the COPY Statement

### DISCUSSION THREAD

**Discussion by arnoldwakim** [duckdb/duckdb#10100] [Q&A]
Hello, 

I'm trying to use DuckDB amazing postgres_scanner to export some data to s3.

Let's say I use the ATTACH semantics to attach a postgres database using the following. 

```sql
"ATTACH 'postgresql://postgres:postgres@postgres:5432/postgres' AS my_db (TYPE postgres, READ_ONLY);"
```

and later perform some query

```sql
COPY (
    SELECT
        d.some_id,
        d.some_other_column
    FROM
        postgres_scan_pushdown($1, 'my_schema', 'my_table') as d
    WHERE
        group_id = $2
) TO 'XXXXXXXXXXXX' (FORMAT 'parquet');
"""
```

Using the python API, I use the following

```python
conn.execute(THE_ABOVE_QUERY, [CONFIG.duck_db_pg_dsn, group_id])
```

I want to replace XXXXXXXXXXXX by some s3 uri which depends on the `group_id`.
Unless I'm mistaken (in that case I apologize in advance) but it looks like it is not possible to  do it with a third placeholder. `$3`.

Could anyone enlighten me on this? 

Thank you for your time,

Best,

Arnold

**Comment by ghost** [comment]
This hit me as well. I think the reason why it can't work with variables/placeholders is that both the "source" and the "target" of the `COPY` statement as considered SQL by themselves and hence, use a different level of syntax validation.

The result is, however, that we're actually open to SQL injections here, since we need to "generate" the full SQL string ourselves at this point.

Something that I've seen in the area of templating across languages and tools is that there are multiple types of placeholders/variables that work on different levels. Something that may be a viable option for DuckDB to adopt:
1. `${variable}` - most common - handle values inside the SQL
2. `#{variable}` - uncommon - applied before validating the SQL syntax - may need explicit metadata to ensure expected content, e.g.:
   1. `#{identifier:variable}` must be string and validate to DB identifier rules (e.g. Table, Column, ...)
   2. `#{sql:variable}` must be string and validate to some SQL (may include value variables) - most dangerous option, though
   3. `#{literal:variable}` must be a string and will be safely single-quoted (can be used to specify a file path)

For reference, MyBatis syntax (although the other way around using `#` for values and `$` for SQL building):
* https://mybatis.org/mybatis-3/sqlmap-xml.html#string-substitution

Just an idea, though.

**Comment by ghost** [comment]
@AqemiArnold Could you add a complete example just as an answer for others to see?

**Reply by arnoldwakim** [reply]
I unfortunately don't have a working example. 
I thought that duckdb variables would help but this is not the case. Sorry.
